### ISSUE TITLE
An enumeration value defined to equal another causes the compiler to erroneously report a circular reference

### ISSUE THREAD

**Issue by ericwj** [dotnet/roslyn#7852]
```CSharp
namespace N {
	/// <summary>The evaluation of the constant value for 'CS0110.Default' involves a circular definition</summary>
	public enum CS0110 {
		Default = Value1, // CS0110: <font color="red">Default</font> = Value1,
		Value1,
		Value2
	}
	/// <summary>The switch statement contains multiple cases with the label value 'case CS0110.Value2:'</summary>
	public static class CS0152 {
		public static void Switch(CS0110 value) {
			switch (value) {
				case CS0110.Value1:
					break;
				case CS0110.Value2: // CS0152: <font color="red">case CS0110.Value2:</font>
					break;
			}
		}
	}

	/// <summary>Compiles happily</summary>
	public enum NoCS0110 {
		//Default = Value1,
		Value1,
		Value2
	}
	/// <summary>Compiles happily</summary>
	public static class NoCS0152 {
		public static void Switch(NoCS0110 value) {
			switch (value) {
				case NoCS0110.Value1:
					break;
				case NoCS0110.Value2:
					break;
			}
		}
	}
}
```

**Comment by gafter** [comment]
```cs
    public enum CS0110 {
        Default = Value1, // CS0110: <font color="red">Default</font> = Value1,
        Value1,
        Value2
    }
```

According to the spec, below, this is the same as

```cs
    public enum CS0110 {
        Default = Value1,
        Value1 = Default + 1,
        Value2 = Value1 + 1
    }
```

> ### 14.3 Enum members
The associated value of an enum member is assigned either implicitly or explicitly. If the declaration of the enum member has a constant-expression initializer, the value of that constant expression, implicitly converted to the underlying type of the enum, is the associated value of the enum member. If the declaration of the enum member has no initializer, its associated value is set implicitly, as follows:
- If the enum member is the first enum member declared in the enum type, its associated value is zero.
- Otherwise, the associated value of the enum member is obtained by increasing the associated value of the textually preceding enum member by one. This increased value must be within the range of values that can be represented by the underlying type, otherwise a compile-time error occurs.
> 
...
> 
**The associated value of an enum member may not, directly or indirectly, use the value of its own associated enum member.** ... 

Since this is a circular definition given your code, an error is required.

**Comment by ericwj** [comment]
Hmm, sharp. Though it's about the only place in C# where declaration order is important. It got me puzzled for a while although the fix obviously is to define Default one line later. It is at least a bit alien for C#, isn't it?

One could say that `Value1` is not a constant expression at this point, it's undefined. That'd make for a better error message that'd be instantly obvious to fix, although it may make the compiler look more stupid than it is. Then for another version maybe one line extra in the spec can make this oddity go away? Just evaluate declarations using constant expressions using enumeration values from the same enumeration after all others got their value as described above.
